Start using `block.json` and convert to TS the `Product by Category` block

albarin opened the pull request:
Converts the `Products by Category` block to Typescript and starts using the `block.json` file.

### Testing

#### User Facing Testing
Test that there are no regressions compared with the previous `Products by Category` block version.

1. Create a page and add a `Products by Category` block.
2. Save the page and check it renders correctly the default config in the frontend.
3. Edit the page again and make some changes to the block (hiding some content, changing columns number, ordering, etc.).
4. Save the page and check it renders properly with the new config. 

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

albarin committed:
Start using `block.json` and convert to TS the `Product by Category` block

gigitux commented on assets/js/blocks/product-category/edit-mode.tsx:
```diff
@@ -0,0 +1,104 @@
+export interface EditModeProps extends Props {
+	isEditing: boolean;
+	setIsEditing: ( isEditing: boolean ) => void;
+	changedAttributes: object;
```
Could we use `Partial< Attributes >`, no?

gigitux commented on assets/js/blocks/product-category/edit-mode.tsx:
```diff
@@ -0,0 +1,104 @@
+	isEditing: boolean;
+	setIsEditing: ( isEditing: boolean ) => void;
+	changedAttributes: object;
+	setChangedAttributes: ( changedAttributes: object ) => void;
```
Same here.

gigitux commented on assets/js/blocks/product-category/edit-mode.tsx:
```diff
@@ -0,0 +1,104 @@
+				'woo-gutenberg-products-block'
+			) }
+			<div className="wc-block-products-category__selection">
+				<ProductCategoryControl
```
![image](https://user-images.githubusercontent.com/4463174/179233010-81f4dbd5-b9c5-4e9e-96de-4a7ce8464f45.png)

To avoid introducing new TS errors, maybe we could add a JSDoc and define some props optional. What do you think?

gigitux commented on assets/js/blocks/product-category/edit.tsx:
```diff
@@ -0,0 +1,72 @@
+		! attributes.categories.length
+	);
+
+	const [ changedAttributes, setChangedAttributes ] = useState( {} );
```
Same here, I think that we could use `Partial< Attributes >`.

gigitux commented on assets/js/blocks/product-category/inspector-controls.tsx:
```diff
@@ -0,0 +1,116 @@
+				) }
+				initialOpen={ ! attributes.categories.length && ! isEditing }
+			>
+				<ProductCategoryControl
```
Same here.

gigitux commented on assets/js/blocks/product-category/inspector-controls.tsx:
```diff
@@ -0,0 +1,116 @@
+				<GridLayoutControl
+					columns={ columns }
+					rows={ rows }
+					alignButtons={ alignButtons }
```
I guess that there is an error in the definition of the type of the props for the component `GridLayoutControl`. I think that we can replace [this](https://github.com/woocommerce/woocommerce-blocks/blob/5fc52a593e9ee2c3b6134f16e2e6b67ce2d9bcc0/assets/js/editor-components/grid-layout-control/index.js/#L16-L17) with `boolean`.

gigitux reviewed (changes requested):
Great work! Everything works! 
I added some comments with the aim of not introducing any TS errors in the codebase. Let's see if it is possible 💪

albarin committed:
Address feedback to remove some TS errors

albarin committed:
Merge branch 'trunk' into product-by-cat-to-ts

albarin commented:
> Great work! Everything works! I added some comments with the aim of not introducing any TS errors in the codebase. Let's see if it is possible 💪

I think I have addressed all your comments, could you have another quick look? 🙏 Thanks!

albarin committed:
Remove unnecessary prop

gigitux reviewed (approved):
Thanks for addressing all the comments! LGTM! 🚢
